### ISSUE TITLE
[hebao] Make the wallet implementation contract upgradeable

### ISSUE THREAD

**Issue by Brechtpd** [Loopring/protocols#808]
Currently the wallet implementation cannot be upgraded by anyone. You could say this is the most secure, but this also has risks (bugs cannot be fixed, gas costs could change that make the smart wallet less usable like #807 which cannot be predicted,...).

I think it would be a good idea to allow the smart wallet implementation to be updated to an approved contract + `(n + 1)/2` signatures that do the actual upgrade.

**Comment by dong77** [comment]
The idea is to allow upgrading a wallet by upgrading its modules with  `UpgraderModule`, and upgrade is ONLY possible by the wallet's owner for the utmost security.

**Comment by Brechtpd** [comment]
How would upgrading the wallet implementation be different from adding/removing modules from a security point of view? It seems like both have about the same amount of risk because modules can do anything on the wallet (of course it's more secure because less code is changed, but that's beside the point).

In both cases the module/wallet implementation would need to be whitelisted. In both cases we can allow only the wallet owner to do it (I chose `(n + 1)/2` for extra security, but the wallet owner would of course also need to be a part).

I understand everyone wants to be in complete control of their wallet, that's why I chose NOT to do the upgrade automatically. We would simply whitelist the newer version of the smart wallet implementation, and then it would be possible (if the wallet owner so decides!) to upgrade to it. All power remains in the hand of the wallet owner, but at least there's a mechanism to upgrade the wallet contract so we don't have problems like [Aragon where they can't fix their smart wallet and people can lose money](https://blog.aragon.org/istanbul-hard-fork-impact/).
